draft-normington-jsonpath-latest.xml

```diff
@@ -384,50 +384,115 @@ array-slice = [ integer ] ws %x3A ws [ integer ]
-                        The indices of the slice are the same as the values of i iterated over by this <tt>for</tt> loop in C:
-                     </t>
-                     <sourcecode type="c">
-for (i = start; i &gt; end; i = i + step) {
-  ...
-}
+
+                     <t>Slice expression parameters <tt>start</tt> and <tt>end</tt> are not directly usable
+                        as slice bounds and must be first normalized. Normalization is defined as:</t>
+
+                     <sourcecode type="pseudocode">
+FUNCTION Normalize(i):
+  IF i &gt;= 0 THEN
+    RETURN i
+  ELSE
+    RETURN len + i
+  END IF
                      </sourcecode>
-                     <t>When <tt>step</tt> is negative, the default value of <tt>start</tt> is <tt>len - 1</tt> (where <tt>len</tt> is the length of the array)
-                        and the default value of <tt>end</tt> is <tt>-1</tt> (literally, not in the earlier sense of a shorthand for <tt>len - 1</tt>).
-                     </t>
-                     <t>For zero <tt>step</tt>, the array slice is empty and no elements are selected.
-                     </t>
+
+                     <t>The semantics of the array indexing expression <tt>[i]</tt> is defined as syntactic sugar for
+                        <tt>a[i:Normalize(i)+1:1]</tt>.</t>
+
+                     <t>Slice expression parameters <tt>start</tt> and <tt>end</tt> are used to derive slice bounds
+                        <tt>lower</tt> and <tt>upper</tt>. The direction of the iteration defined
+                        by the sign of <tt>step</tt> determines which of the parameters is the lower bound and which
+                        is the upper bound:</t>
+
+                     <sourcecode type="pseudocode">
+FUNCTION Bounds(start, end, step, len):
+  n_start = Normalize(start)
+  n_end = Normalize(end)
+
+  IF step &gt;= 0 THEN
+    lower = MIN(MAX(n_start, 0), len)
+    upper = MIN(MAX(n_end, 0), len)
+  ELSE
+    upper = MIN(MAX(n_start, -1), len-1)
+    lower = MIN(MAX(n_end, -1), len-1)
+  END IF
+
+  RETURN (lower, upper)
+</sourcecode>
+
+                     <t>The slice expression selects indices between the lower and upper bounds.
+                        In the following pseudocode, the <tt>a(i)</tt> construct expresses the
+                        0-based indexing operation on the underlying array.</t>
+
+                     <sourcecode type="pseudocode">
+IF step &gt; 0 THEN
+
+  i = lower
+  WHILE i &lt; upper:
+    SELECT a(i)
+    i = i + step
+  END WHILE
+
+ELSE if step &lt; 0 THEN
+
+  i = upper
+  WHILE lower &lt; i:
+    SELECT a(i)
+    i = i + step
+  END WHILE
+
+END IF
+                     </sourcecode>
+
+                     <t>The case when <tt>step = 0</tt> intentionally doesn't select any array element
+                        and thus produces an empty result array.</t>
+
+                     <t>All parameters, including the array length, SHOULD be representable as a signed 64-bit integer.
```

Comment:
I'm not sure what value this statement adds and it could even be viewed as implementation bias. The following statements say what to do when a value is encountered which the implementation can't cope with.

Reply:
Makes sense, thanks

Reply:
To be clear, I was suggesting deleting the sentence.

Reply:
suggestion applied